langchain4j-core/src/main/java/dev/langchain4j/model/chat/StreamingChatLanguageModel.java

```diff
@@ -1,26 +1,105 @@
+import dev.langchain4j.model.chat.request.ChatRequest;
+import dev.langchain4j.model.chat.request.ResponseFormat;
+import dev.langchain4j.model.chat.request.ResponseFormatType;
+import dev.langchain4j.model.chat.response.ChatResponse;
+import dev.langchain4j.model.chat.response.StreamingChatResponseHandler;
+import dev.langchain4j.model.output.Response;
 
 import java.util.List;
 
+import static dev.langchain4j.internal.Utils.isNullOrEmpty;
 import static java.util.Collections.singletonList;
 
 /**
+ * TODO review all javadoc in this class
  * Represents a language model that has a chat interface and can stream a response one token at a time.
+ *
+ * @see ChatLanguageModel
  */
 public interface StreamingChatLanguageModel {
 
+    /**
+     * TODO
+     * <p>
+     * A temporary default implementation of this method is necessary
+     * until all {@link StreamingChatLanguageModel} implementations adopt it. It should be removed once that occurs.
+     *
+     * @param chatRequest
+     * @param handler
+     */
+    @Experimental
+    default void chat(ChatRequest chatRequest, StreamingChatResponseHandler handler) {
+
+        ResponseFormat responseFormat = chatRequest.responseFormat();
+        if (responseFormat != null && responseFormat.type() == ResponseFormatType.JSON) {
+            // TODO check supportedCapabilities() instead?
+            throw new UnsupportedOperationException("JSON response type is not supported by this model provider");
+        }
+
+        StreamingResponseHandler<AiMessage> legacyHandler = new StreamingResponseHandler<>() {
+
+            @Override
+            public void onNext(String token) {
+                handler.onPartialResponse(token);
+            }
+
+            @Override
+            public void onComplete(Response<AiMessage> response) {
+                ChatResponse chatResponse = ChatResponse.builder()
+                        .aiMessage(response.content())
+                        .tokenUsage(response.tokenUsage())
+                        .finishReason(response.finishReason())
+                        .build();
+                handler.onCompleteResponse(chatResponse);
+            }
+
+            @Override
+            public void onError(Throwable error) {
+                handler.onError(error);
+            }
+        };
+
+        if (isNullOrEmpty(chatRequest.toolSpecifications())) {
+            generate(chatRequest.messages(), legacyHandler);
+        } else {
+            generate(chatRequest.messages(), chatRequest.toolSpecifications(), legacyHandler);
+        }
+    }
+
+    /**
+     * TODO
+     * <p>
+     * A temporary default implementation of this method is necessary
+     * until all {@link StreamingChatLanguageModel} implementations adopt it. It should be removed once that occurs.
+     *
+     * @param userMessage
+     * @param handler
+     */
+    @Experimental
+    default void chat(String userMessage, StreamingChatResponseHandler handler) {
```

Comment:
@glaforge what do you think about this API?
User will be able to pass `Consumer<String> partialResponseHandler` as lambda and also have an option to wait till the completion of full request or error (via `CompletableFuture`)

Reply:
Hmm interesting idea, and maybe more modern.
Would this be just for the `String` variant? I suppose that would be also for the method taking the `ChatRequest` parameter?
And doing so, we would not need the `onComplete()` and `onError()` methods at all?

Reply:
I am not sure if this should be for convenience method or for both.

The reasoning for convenience method: user just wants to test the model, so it should be easy and fast to do that, it does not have to force dealing with errors and complete response by default.
So one can just write this: `model.chat("hello", System.out::println);` and it will work.

And only if one wants to deal with complete response or errors, one can do:
```java
CompletableFuture<ChatResponse> future = model.chat("hello", System.out::println);
ChatResponse chatResponse = future.get(10, TimeUnit.SECONDS);
```

I am not so sure about the `chat(ChatRequest)` API, as it should be future-proof.
In case LLMs will start streaming something other than text (e.g., images or audio) we will have to again add more methods for each case in `StreamingChatLanguageModel`:
- `chat(ChatRequest, Consumer<Image>)`
- `chat(ChatRequest, Consumer<Audio>)`

I guess it will be easier to add new default methods in `StreamingChatResponseHandler` instead.

WDYT?

Reply:
I assume that in most cases users will use AI Service and `TokenStream`/`Flux<String>`/etc return type, so I guess the priorities for the low-level `StreamingChatLanguageModel` API should go in this order:
1. flexible enough to accomodate new features
2. easy to use

Reply:
Regarding `chat(ChatRequest, Consumer<Image>)`... couldn't it be `chat(ChatRequest, Consumer<Content>)` maybe? (I think there's an ongoing PR about using interfaces verywhere)

Reply:
Perharps, but I am not sure we should introduce it now "just in case". `chat(ChatRequest, Consumer<Image>)` is a hypothetical example. What I meant is that with callback interface, it is more flexible for extension in the furture.

TBH this takes a lot more time than I expected to spend here. And there is still no clear conclusion. Let's just remove that helper method and move forward. We can always add it in the future (if there will be a real need). This is a low-level API, is it not meant to be the easiest to use (for that we have AI Services), so having a single chat() method is fine from my point of view.